internal/pkg/task/task.go

```diff
@@ -0,0 +1,112 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package task provides methods to run task in different configurations.
+package task
+
+import (
+	"errors"
+	"fmt"
+	"github.com/aws/copilot-cli/internal/pkg/aws/ec2"
+	"github.com/aws/copilot-cli/internal/pkg/aws/ecs"
+	"github.com/aws/copilot-cli/internal/pkg/aws/resourcegroups"
+	"github.com/aws/copilot-cli/internal/pkg/aws/session"
+)
+
+type vpcGetter interface {
+	GetSubnetIDs(filters ...ec2.Filter) ([]string, error)
+	GetSecurityGroups(filters ...ec2.Filter) ([]string, error)
+}
+
+type resourceGetter interface {
+	GetResourcesByTags(resourceType string, tags map[string]string) ([]*resourcegroups.Resource, error)
+}
+
+type defaultClusterGetter interface {
+	DefaultCluster() (string, error)
+}
+
+type taskRunner interface {
+	RunTask(input ecs.RunTaskInput) ([]string, error)
+}
+
+const (
+	startedBy           = "copilot-task"
+)
+
+var (
+	errNoSubnetFound     = errors.New("no subnets found")
+	fmtTaskFamilyName    = "copilot-%s"
+)
+
+// Task gets information required to run a task.
+type Task struct {
+	count     int
+	groupName string
+
+	vpcGetter      vpcGetter
+	resourceGetter resourceGetter
+	clusterGetter  defaultClusterGetter
+	starter        taskRunner
+}
+
+// NewTaskConfig contains fields required to initialize a task struct.
+type NewTaskConfig struct {
+	// Count is the count of the tasks to be run
+	Count int
+	// GroupName is the name of the task group
+	GroupName string
+}
+
+// NewTask initializes a new task struct.
+func NewTask(config NewTaskConfig) (*Task, error) {
```

Comment:
I'm thinking we should consider stopping with constructors whenever we can that initializes sessions and then our dependencies. Instead, we can do dependency injection from the "cli" pkg.

> I’ve found that just writing code and just instantiating things, and just passing them in inside of main tends to be the best, most straightforward dependency injection I’ve seen. That’s all I do.
https://changelog.com/gotime/102

So the fields in the `Runner` struct would be public as well as the interfaces in this package. So the `cli` pkg is the only package that creates a session and the aws clients, then passes them down to the `task.Runner` struct and all of our other structs.


Does that make sense to you? What do you think?

Reply:
Yeah I can see the merit of it. That'd mean if I want to use `Runner`, I'd need to be aware of its dependencies. Is that correct? Would that result in too much complexity in the caller, e.g. say `cli` package (having to know all the dependencies of its dependencies)?

Reply:
Yeah good question, I agree that in this situation the CLI needs to ensure that all the dependencies are set that's why I am also torn lol  

I think the right answer is probably copying what the standard library does. For example, the [http.Server](https://golang.org/pkg/net/http/#Server) struct has lots of configuration similar to ours. However, if a user does not provide a handler than they set default ones (https://golang.org/src/net/http/server.go?#L2800) or panic. Maybe in our situation, we can check if they're nil and return an error 🤔

Reply:
Sounds good. The merit of returning an error over having a default behavior is that the behavior of, say `Runner` is more predictable right? Because the caller knows what the fields are set to, otherwise they would have received an error?

I have another question about having all fields public. Why would that be necessary in the case of dependency injection? Do we want to allow the caller to be able to later alter the fields after the instance is initiated?

Reply:
> Because the caller knows what the fields are set to, otherwise they would have received an error?

I think the std lib panics if there is no good default behavior that can be set.   

OK how about this, we don't do any validation within `Run` and assume that these fields need to be set. We can add a comment to make that explicit.
To guarantee that our code handles that, we can add an integ test to the `cli` pkg to makes sure the `task run` command works and initializes the dependencies properly.

> I have another question about having all fields public. Why would that be necessary in the case of dependency injection? 

I might be misunderstanding but how else could the `cli` package set the field for the `task.DefaultVPCConfig` struct if it's not public? I meant that they should be public so that the clients can set them (not to alter)